### ISSUE TITLE
Support externally hosted Grafana dashboards

### ISSUE THREAD

**Issue by jackgill** [linkerd/linkerd2#7429]
### What problem are you trying to solve?
I would like to host Linkerd's Grafana dashboards using a managed service such as Grafana Cloud or Amazon Managed Grafana. However, dashboard URL's for these services have a different format than the on-cluster Grafana instance. Currently Linkerd's Grafana dashboard links are created using this string template:
https://github.com/linkerd/linkerd2/blob/8ddba7a27e77f872a3e84f2a3b794c4b418dcd80/web/app/js/components/GrafanaLink.jsx#L7
However, for externally hosted Grafana the template would need to look like this:
```golang
let link = `/d/linkerd-${resource}/linkerd-${resource}?var-${resource}=${name}`;
```
Note the absence of a `/grafana/` URL segment and the addition of a URL segment for a user-specified UID, which in this example I have chosen to be the same as the dashboard name.

There is a potentially related issue (#7358) regarding how the dashboard definitions are managed, but as far as I can tell it still assumes that the Grafana instance is on-cluster and follows the existing URL conventions.

Also, the dashboard definitions have a hard-coded data source name and include some Javascript and image links which are disallowed by some third-party hosts.

### How should the problem be solved?
Since different hosting providers may have different URL formats, I am thinking that the entire path template should be user-configurable, with a default value of the current template to maintain backwards compatibility.

There is already a `grafanaURL` variable, but it is only used to build the scheme and host parts of the URL. I suggest adding a new variable called `grafanaPathTemplate`. Although I am new to this project I would be happy to contribute code for this if the maintainers think it's a good idea.

We could also do some templating with the dashboard definitions to add a `grafanaDataSource` variable, and a flag to omit Javascript and image links. However, these issues can be addressed without modifying Linkerd so I consider them a lower priority.

### Any alternatives you've considered?
I think this could also be addressed by running a proxy that rewrites URL's, but that seems like an unnecessary amount of overhead.

### How would users interact with this feature?
By optionally specifying a value for `grafanaPathTemplate` in `values.yaml`.

**Comment by alpeb** [comment]
That's great feedback, thanks @jackgill.

I think I agree on most of your points, so let me state the required changes again perhaps with slight relabeling :wink: :

Indeed, `grafanaUrl` only defines a location `internal` to the cluster. If empty it defaults to something like `grafana.linkerd-viz.svc.cluster.local:3000`. And the dashboard web server makes a port-forward pointing to that, for any url starting with `/grafana`. We could add a new setting `grafanaExternalUrl` (defaulting to empty) with the full link to the external Grafana instance, passed to the web deployment which in turns passes it to index.js, just like we do with `grafana-addr`. And then have `GrafanaLink.jsx` override the default link with this new one if set.

And yes, It would also be cool to have another setting for parametrizing the data source name in the templates and another to disable the javascript images. 

#7358 is only slightly related, as it deals about how to declare the grafana dashboards definitions to be used by a non-linkerd Grafana instance. If that instance is gonna be external then you need to configure it to point to the published dashboards (option 1 in that issue), or render their ConfigMaps (option 3, properly replacing the parameters mentioned above) so you can copy-paste them. Does this make sense?

Anyways, we would very much welcome your contribution if you decide to give it a shot, and I'd be glad to provide guidance :+1:

**Comment by jackgill** [comment]
@alpeb thank you for the relabeling, that is indeed what I would like to accomplish. And I would certainly appreciate your guidance! Perhaps I could take a rough cut at a PR and then you could provide some feedback? Or if there is anything you think I should know before starting, let me know.

**Comment by alpeb** [comment]
@jackgill are you still interested on tackling this one? :slightly_smiling_face:

**Comment by jackgill** [comment]
@alpeb yes, I should have a PR for you to look at within the next couple days.
